### ISSUE TITLE
Absence of non-Windows FlashTool is bottleneck in class

### ISSUE THREAD

**Issue by wdconinc** [nickcounts/Arduino-Lessons#37]
Many students have Macs (or even chromebooks), and can't use FlashTool to flash the boards initially. The radio firmware for CR03 is provided as hex files as well, which allows them to flash those boards on Macs or other systems. Suggestion: the firmware that FlashTool uses should be made available as a hex file as well.

**Comment by nickcounts** [comment]
I agree, the lack of non-windows tools should be addressed. (Who'd have thought we'd see the day when windows only was a problem!)

I believe a future version of the FlashTool will support the loading of an external binary for flashing onto the CW01. I am not sure how XinaBox is planning to distribute the binaries, but I assume they will be on their GitHub repo (although possibly served up from their server to the tool).

Since the program was envisioned as an outreach program for middle school and high school, I believe this functionality isn't a high priority. As a Mac/Linux user, I feel your pain.

I'll check in and see what we can get for you. In the meantime, perhaps we can both investigate pulling the binary out of the FlashTool.exe resource bundle? That may allow you to use other programming software until XinaBox and ThinSat roll out new multi-platform tools.

**Comment by nickcounts** [comment]
I dug around in the FlashTool executable and found two binaries. I haven't tested these yet, but I posted them to our repository so you can take a look

[xinaboxFlashTool_X_XINABOX_BIN.bin](https://github.com/VCSFA-MARS/ThinSat-Program/commit/3bd05db602195ab108f241cf38abf056dedb0e60)

The mircopython binary is also available.

Let us know if you have any luck with these

**Comment by wdconinc** [comment]
I'm assuming something like
```
avrdude -p m328p -b 57600 -P /dev/ttyUSB0 -c stk500v1 -U flash:w:xinaboxFlashTool__X_XINABOX_BIN.bin
```
would do the job. I don't have a board on me, but I'll see tomorrow if that works.

**Comment by wdconinc** [comment]
I was unable to flash using avrdude above (seems at least baud rate should be 115200). But I was now able to get it to flash with FlashTool.exe under wine. What was needed (COM3 to `/dev/ttyUSB0`):
1. Create symlink to setup COM3 to `/dev/ttyUSB0` connection: `ln -sf /dev/ttyUSB0 ~/.wine/dosdevices/com3`. You will need to be in a group that has permissions to access `/dev/ttyUSB0`, likely `dialout` or so.
2. Modify registry with `wine regedit` and add key `HKEY_LOCAL_MACHINE\Software\Wine\Ports` with string entry `COM3` with value `/dev/ttyUSB0`.
3. Reset wine server: `wineserver -k`.
4. Start FlashTool in wine: `wine xinaboxFlashTool.exe`.
5. Select `XINABOX` firmware and `Safe Flash` (it didn't work without safe flash for me).
6. Type in `COM3` in serial port dropdown (nothing in dropdown).
7. Click `Flash`.
8. Toggle DTE/DCE until it detects when going to DCE.

![image](https://user-images.githubusercontent.com/4656391/32911095-2013c132-cad9-11e7-9cca-b90b120dc734.png)

**Comment by nickcounts** [comment]
Nice fix on Wine - I was having issues on mine and didn't think to edit the registry. I'll give it a shot.

I saw that you made a PR to XinaBox and wanted to add that avrdude is available on for OS X with homebrew: `brew install avrdude`

**Comment by EBUJOLD** [comment]
This is something that should be directed to the XinaBox team. They developed the FlashTool and we have discussed with them ways to improve the tool. Maybe if you reach out to them with more ideas, that would be valuable. 

https://github.com/xinabox/FlashTool

**Comment by EBUJOLD** [comment]
@wdconinc Have you been able to flash the chip with AVRDUDE since your last attempt?

**Comment by wdconinc** [comment]
Not tried again after I got wine to work. I do think I should have more luck with stk500v1 now. I'll try again tomorrow!

**Comment by nickcounts** [comment]
I had success flashing the CW01 from my OS X command line using `esptool` (a python utility)and the XinaBox IP01 USB Interface with the switches set in `B` and `DCE`:

We have added the tested/working version of this tool as a submodule to the ThinSat repo.


# My Commands
```
git clone https://github.com/themadinventor/esptool.git
cd esptool
sudo python setup.py install
sudo python esptool.py --baud 115200 --port /dev/tty.usbserial-AI058C59 write_flash -fm dio 0x00000 /Users/nickcounts/Documents/Repositories/ThinSat-Program/compiledBinaries/flashTool-v-1.207/xinaboxFlashTool__X_XINABOX_BIN.bin 
```

# Console Output

```
esptool.py v2.2
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0240
Compressed 402160 bytes to 282598...
Wrote 402160 bytes (282598 compressed) at 0x00000000 in 25.1 seconds (effective 128.1 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting...
```

![img_4599](https://user-images.githubusercontent.com/10506020/33509851-4614a18e-d6d3-11e7-929c-b3cd90912c76.JPG)


Using the binaries attached in our repo (referenced above), this should absolutely work. I believe it will work on linux distros as well (I don't have my linux box on me, so I can't test right now).

**Comment by brendanvanbreda** [comment]
@wdconinc  Please have a look at [xFlasher](https://github.com/xinabox/xFlasher) this should solve all your problems.
